### ISSUE TITLE
Upload Built Wheels to PyPI

### ISSUE THREAD

**Issue by seanlaw** [louisabraham/pydivsufsort#37]
Currently, we have a [Github Actions Workflow](https://github.com/louisabraham/pydivsufsort/blob/master/.github/workflows/github-actions.yml) building all of the appropriate wheels and a source distribution (sdist). However, they still need to be uploaded to PyPI via Twine (don't forget to bump the version in `setup.py`!!).

**Comment by seanlaw** [comment]
@louisabraham I've done all that I can with the Github Actions migration. Would you mind handling this part? Feel free to ask any clarifying questions!

**Comment by louisabraham** [comment]
Sure, will do.

**Comment by seanlaw** [comment]
@louisabraham Currently, the Github Actions will trigger every time new code is committed to the repo OR if there is a PR. You'll probably want to consider possibly creating different Github Actions "workflows" for PRs versus a release.

**Comment by louisabraham** [comment]
Ok I think I did it. I created lighter workflow for most purposes. `build-and-upload` is only launched when a commit modifying `setup.py` (where the version is specified) is pushed to master. We could even read the version from a separate file to isolate version changes even more.

**Comment by seanlaw** [comment]
> Ok I think I did it. I created lighter workflow for most purposes. build-and-upload is only launched when a commit modifying setup.py (where the version is specified) is pushed to master. We could even read the version from a separate file to isolate version changes even more.

Awesome! Thanks for working on it. From what I've gleaned from other projects, it is common to trigger `build-and-upload` based on the commit message OR if the commit is tagged: 

https://stackoverflow.com/questions/61891328/trigger-github-action-only-on-new-tags

So, `build-and-upload` doesn't trigger every time and only if you manually tag a commit with, say, "Release version 0.06" (and then you'd match for "Release version " or something like that)

**Comment by louisabraham** [comment]
I prefer my solution of watching `setup.py`. Also, there is no risk to overwrite a version.

One run is not ending: https://github.com/louisabraham/pydivsufsort/runs/4499216537

**Comment by louisabraham** [comment]
I also fixed the code coverage: https://app.codecov.io/gh/louisabraham/pydivsufsort

**Comment by seanlaw** [comment]
> I prefer my solution of watching `setup.py`

I see. So you are allowing the workflow to trigger only on `push` and when `setup.py` is modified (I guess only you and I have permission to do this). And then you have the `test` workflow for PRs. Cool!

**Comment by louisabraham** [comment]
It also has to run on the master branch, and only we two can push to master. I relaunched the job that didn't run to completion, this time I'll not stop it (github might after 6 hours).

It doesn't seem stuck, just very slow: https://github.com/louisabraham/pydivsufsort/runs/4499663161?check_suite_focus=true

**Comment by seanlaw** [comment]
> I think it's because of our numpy requirement as there are no wheels uploaded for that old version and we must recompile numpy ourselves for each version.

But what actually changed from what my PR to now? All of my wheels were built successfully.

**Comment by louisabraham** [comment]
Oh, good point. Nothing changed. Your PR built all the wheels in 2h30 so I should expect the same timing: https://github.com/louisabraham/pydivsufsort/runs/4168448448?check_suite_focus=true

I think my diagnosis is right. Search the [logs](https://pipelines.actions.githubusercontent.com/wYjIygmbNAPjpDc7I9CMHQe0PeFc2KVGv5ciKHFiN41s4a8NO3/_apis/pipelines/1/runs/54/signedlogcontent/126?urlExpires=2021-12-12T22%3A58%3A53.4243958Z&urlSigningMethod=HMACV1&urlSignature=V2kJvem4nhXknb2AndCEz9Xcv4Dyfx0PiMRv63wYMw4%3D) for "Building wheels for collected packages: numpy". It's really this step that makes one build go from 1 minute to 15 minutes.

On some platforms we have to build numpy twice: once for building (`oldest-supported-numpy`) and one for testing (`numpy`). I think this could be improved by asking the numpy folks to push wheels.

**Comment by louisabraham** [comment]
see https://github.com/numpy/numpy/issues/20573

**Comment by louisabraham** [comment]
it finished!
What a nice sight :)
<img width="539" alt="image" src="https://user-images.githubusercontent.com/13174805/145733742-05a511f2-c683-49ab-8819-b30e539f27f7.png">

**Comment by seanlaw** [comment]
That's great! Thanks again for pushing it through the final meter!

**Comment by louisabraham** [comment]
I excluded musl from the builds since people need to build numpy anyway for the installation.
Also, upgrading pip allows it to find wheels for `cp310-musllinux_x86_64`.
This reduced the build time from 2h30 to 15 min.

I kept `cp310-manylinux_i686` even though we have to build `oldest-supported-numpy` because people can install it without compilation since there are wheels for `numpy`, it takes about 5 minutes.
